File: Client/JsCache.hs
{-# LANGUAGE OverloadedStrings #-}

-- JsCache is used to concatenate Yoink modules

import qualified System.Environment as Env
import qualified Data.Text          as Text
import qualified Data.Text.IO       as TextIO


main :: IO ()
main = do
    args <- Env.getArgs
    let (outnm, files) = parseArgs args

    strs <- mapM TextIO.readFile files
    let cnts = preloadedMods (zip files strs)

    case outnm of
        Nothing -> putStr (Text.unpack cnts)
        Just nm -> TextIO.writeFile nm cnts

parseArgs :: [String] -> (Maybe FilePath, [FilePath])
parseArgs ("-o":p:xs) = (Just p, xs)
parseArgs xs          = (Nothing, xs)


preloadedMods :: [(FilePath, Text.Text)] -> Text.Text
preloadedMods xs = Text.concat [
    "var PRELOADED_MODULES = {\n",
    Text.intercalate ",\n" (map wrapFile xs),
    "};\n"
    ]

wrapFile :: (FilePath, Text.Text) -> Text.Text
wrapFile (p, cnts) = Text.concat [
    "\"/",
    Text.pack p,
    "\": (function (baseUrl, define, require, params) {\"use strict\";\n",
    cnts,
    "\n})"
    ]

File: Client/JsCache.lua
-- JsCache is used to concatenate Yoink modules

local function readfile(p)
    local f = assert(io.open(p, 'r'))
    local s = f:read('*all');
    f:close()
    return s
end

-- Get output file
local out
if arg[1] == '-o' then
    local outfile = arg[2]
    out = assert( io.open(outfile, 'w') )

    table.remove(arg, 1)  -- pop
    table.remove(arg, 1)  -- pop
else
    out = io.stdout
end

-- The rest of the arguments are files
local files = arg

out:write('var PRELOADED_MODULES = {\n')
for i,path in ipairs(files) do
    local text = readfile(path)
    out:write('"/' .. path .. '": ')
    out:write('(function (Yoink) {"use strict";\n')
    out:write(text)
    out:write('\n})')

    if i ~= #files then
        out:write(',\n')
    end
end
out:write('};\n')

out:close()
   
